osquery/tables/system/linux/md_tables.cpp

```diff
@@ -0,0 +1,735 @@
+  auto pieces(split(line, " "));
+  if (pieces.size() != 4) {
+    LOG(WARNING) << "Unexpected recovery/resync line format: " << line;
+    return;
+  }
+  trimStrs(pieces);
+
+  result.progress = pieces[0] + " " + pieces[1];
+
+  auto start = pieces[2].find_first_not_of("finish=");
+  if (start != std::string::npos) {
+    result.finish = pieces[2].substr(start);
+
+  } else {
+    result.finish = pieces[2];
+  }
+
+  start = pieces[3].find_first_not_of("speed=");
+  if (start != std::string::npos) {
+    result.speed = pieces[3].substr(start);
+
+  } else {
+    result.speed = pieces[3];
+  }
+}
+
+void parseMDBitmap(std::string& line, MDBitmap& result) {
+  auto bitmapInfos(split(line, ","));
+  if (bitmapInfos.size() < 2) {
+    LOG(WARNING) << "Unexpected bitmap line structure: " << line;
+
+  } else {
+    trimStrs(bitmapInfos);
+    result.onMem = bitmapInfos[0];
+    result.chunkSize = bitmapInfos[1];
+
+    if (bitmapInfos.size() > 2) {
+      auto pos = bitmapInfos[2].find("file:");
+      if (pos != std::string::npos) {
+        result.externalFile = bitmapInfos[2].substr(pos + sizeof("file:") - 1);
+        boost::algorithm::trim(result.externalFile);
+      }
+    }
+  }
+}
+
+MDDrive parseMDDrive(std::string& name) {
+  MDDrive drive;
+  drive.name = name;
+
+  auto start = name.find('[');
+  auto end = name.find(']');
+  if (start == std::string::npos || end == std::string::npos) {
+    LOG(WARNING) << "Unexpected drive name format: " << name;
+    return drive;
+  }
+
+  drive.pos = std::stoi(name.substr(start + 1, end - start - 1));
+
+  return drive;
+}
+
+void MD::parseMDStat(std::vector<std::string>& lines, MDStat& result) {
+  // Will be used to determine starting point of lines to work on.
+  size_t n = 0;
+
+  if (lines.size() < 1) {
+    return;
+  }
+
+  // This should always evaluate to true, but just in case we check.
+  if (lines[0].find("Personalities :") != std::string::npos) {
+    std::string pline(lines[0].substr(sizeof("Personalities :") - 1));
+    parseMDPersonalities(pline, result.personalities);
+
+    n = 1;
+
+  } else {
+    LOG(WARNING) << "mdstat Personalites not found at line 0: " << lines[0];
+  }
+
+  while (n < lines.size()) {
+    if (lines[n].find_first_not_of("\t\r\v ") == std::string::npos) {
+      n += 1;
+      continue;
+    }
+    // Work off of first 2 character instead of just the first to be safe.
+    std::string firstTwo = lines[n].substr(0, 2);
+    if (firstTwo == "md") {
+      auto mdline(split(lines[n], ":", 1));
+      if (mdline.size() < 2) {
+        LOG(WARNING) << "Unexpected md device line structure: " << lines[n];
+        continue;
+      }
+
+      MDDevice mdd;
+      mdd.name = mdline[0];
+      boost::algorithm::trim(mdd.name);
+
+      auto settings(split(mdline[1], " "));
+      trimStrs(settings);
+      // First 2 of settings are always status and RAID level
+      if (settings.size() >= 2) {
+        mdd.status = settings[0];
```

Comment:
Could move the content of `settings` here, then below with `parseMDDrive` you can use `mdd.raidlevel`.